File: SignalServiceKit/src/Storage/AxolotlStore/CallKitIdStore.m
//
// Copyright 2019 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "CallKitIdStore.h"
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

@implementation CallKitIdStore

+ (SDSKeyValueStore *)phoneNumberStore
{
    return [[SDSKeyValueStore alloc] initWithCollection:@"TSStorageManagerCallKitIdToPhoneNumberCollection"];
}

+ (SDSKeyValueStore *)serviceIdStore
{
    return [[SDSKeyValueStore alloc] initWithCollection:@"TSStorageManagerCallKitIdToUUIDCollection"];
}

+ (SDSKeyValueStore *)groupIdStore
{
    return [[SDSKeyValueStore alloc] initWithCollection:@"TSStorageManagerCallKitIdToGroupId"];
}

#pragma mark -

+ (void)setThread:(TSThread *)thread forCallKitId:(NSString *)callKitId
{
    OWSAssertDebug(callKitId.length > 0);

    DatabaseStorageWrite(self.databaseStorage, ^(SDSAnyWriteTransaction *transaction) {
        TSGroupModel *groupModel = [thread groupModelIfGroupThread];
        if (groupModel) {
            [self.groupIdStore setData:groupModel.groupId key:callKitId transaction:transaction];
            // This is probably overkill since we currently generate these IDs randomly,
            // but better futureproof than sorry.
            [self.serviceIdStore removeValueForKey:callKitId transaction:transaction];
            [self.phoneNumberStore removeValueForKey:callKitId transaction:transaction];
        } else {
            OWSAssertDebug([thread isKindOfClass:[TSContactThread class]]);
            SignalServiceAddress *address = [(TSContactThread *)thread contactAddress];
            NSString *serviceIdString = address.serviceIdUppercaseString;
            if (serviceIdString) {
                [self.serviceIdStore setString:serviceIdString key:callKitId transaction:transaction];
                [self.phoneNumberStore removeValueForKey:callKitId transaction:transaction];
            } else {
                OWSFailDebug(@"making a call to an address with no UUID: %@", address.phoneNumber);
                [self.phoneNumberStore setString:address.phoneNumber key:callKitId transaction:transaction];
                [self.serviceIdStore removeValueForKey:callKitId transaction:transaction];
            }

            [self.groupIdStore removeValueForKey:callKitId transaction:transaction];
        }
    });
}

+ (nullable TSThread *)threadForCallKitId:(NSString *)callKitId
{
    OWSAssertDebug(callKitId.length > 0);

    __block TSThread *_Nullable result;
    [self.databaseStorage readWithBlock:^(SDSAnyReadTransaction *transaction) {
        // Most likely: modern 1:1 calls
        NSString *_Nullable serviceIdString = [self.serviceIdStore getString:callKitId transaction:transaction];
        if (serviceIdString) {
            SignalServiceAddress *address = [[SignalServiceAddress alloc] initWithServiceIdString:serviceIdString];
            result = [TSContactThread getThreadWithContactAddress:address transaction:transaction];
            return;
        }

        // Next try group calls
        NSData *_Nullable groupId = [self.groupIdStore getData:callKitId transaction:transaction];
        if (groupId) {
            result = [TSGroupThread fetchWithGroupId:groupId transaction:transaction];
            return;
        }

        // Finally check the phone number store, for very old 1:1 calls.
        NSString *_Nullable phoneNumber = [self.phoneNumberStore getString:callKitId transaction:transaction];
        if (phoneNumber) {
            SignalServiceAddress *address = [[SignalServiceAddress alloc] initWithPhoneNumber:phoneNumber];
            result = [TSContactThread getThreadWithContactAddress:address transaction:transaction];
            return;
        }
    }];

    return result;
}

@end

NS_ASSUME_NONNULL_END

File: Signal/Calls/CallKitIdStore.swift
//
// Copyright 2024 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import LibSignalClient
import SignalRingRTC
import SignalServiceKit
import SignalUI

class CallKitIdStore {
    private static let phoneNumberStore = KeyValueStore(collection: "TSStorageManagerCallKitIdToPhoneNumberCollection")
    private static let serviceIdStore = KeyValueStore(collection: "TSStorageManagerCallKitIdToUUIDCollection")
    private static let groupIdStore = KeyValueStore(collection: "TSStorageManagerCallKitIdToGroupId")
    private static let callLinkStore = KeyValueStore(collection: "CallKitIdToCallLink")

    static func setGroupId(_ groupId: GroupIdentifier, forCallKitId callKitId: String) {
        SSKEnvironment.shared.databaseStorageRef.write { tx in
            // Make sure it doesn't exist, but only in DEBUG builds.
            assert(!phoneNumberStore.hasValue(callKitId, transaction: tx))
            assert(!serviceIdStore.hasValue(callKitId, transaction: tx))
            assert(!groupIdStore.hasValue(callKitId, transaction: tx))
            assert(!callLinkStore.hasValue(callKitId, transaction: tx))

            groupIdStore.setData(groupId.serialize().asData, key: callKitId, transaction: tx)
        }
    }

    static func setContactThread(_ thread: TSContactThread, forCallKitId callKitId: String) {
        SSKEnvironment.shared.databaseStorageRef.write { tx in
            // Make sure it doesn't exist, but only in DEBUG builds.
            assert(!phoneNumberStore.hasValue(callKitId, transaction: tx))
            assert(!serviceIdStore.hasValue(callKitId, transaction: tx))
            assert(!groupIdStore.hasValue(callKitId, transaction: tx))
            assert(!callLinkStore.hasValue(callKitId, transaction: tx))

            let address = thread.contactAddress
            if let serviceIdString = address.serviceIdUppercaseString {
                serviceIdStore.setString(serviceIdString, key: callKitId, transaction: tx)
            } else if let phoneNumber = address.phoneNumber {
                owsFailDebug("making a call to an address with no UUID")
                phoneNumberStore.setString(phoneNumber, key: callKitId, transaction: tx)
            } else {
                owsFailDebug("making a call to an address with no phone number or uuid")
            }
        }
    }

    static func setCallLink(_ callLink: CallLink, forCallKitId callKitId: String) {
        SSKEnvironment.shared.databaseStorageRef.write { tx in
            // Make sure it doesn't exist, but only in DEBUG builds.
            assert(!phoneNumberStore.hasValue(callKitId, transaction: tx))
            assert(!serviceIdStore.hasValue(callKitId, transaction: tx))
            assert(!groupIdStore.hasValue(callKitId, transaction: tx))
            // Call Links may be stored multiple times...

            callLinkStore.setData(callLink.rootKey.bytes, key: callKitId, transaction: tx)
        }
    }

    static func callTarget(forCallKitId callKitId: String) -> CallTarget? {
        return SSKEnvironment.shared.databaseStorageRef.read { tx -> CallTarget? in
            // Most likely: modern 1:1 calls
            if let serviceIdString = serviceIdStore.getString(callKitId, transaction: tx) {
                let address = SignalServiceAddress(serviceIdString: serviceIdString)
                return TSContactThread.getWithContactAddress(address, transaction: tx).map { .individual($0) }
            }

            // Next try group calls
            if
                let groupIdData = groupIdStore.getData(callKitId, transaction: tx),
                let groupId = try? GroupIdentifier(contents: [UInt8](groupIdData))
            {
                return .groupThread(groupId)
            }

            // Check the phone number store, for very old 1:1 calls.
            if let phoneNumber = phoneNumberStore.getString(callKitId, transaction: tx) {
                let address = SignalServiceAddress.legacyAddress(serviceIdString: nil, phoneNumber: phoneNumber)
                return TSContactThread.getWithContactAddress(address, transaction: tx).map { .individual($0) }
            }

            if let rootKeyBytes = callLinkStore.getData(callKitId, transaction: tx) {
                return (try? CallLinkRootKey(rootKeyBytes)).map { .callLink(CallLink(rootKey: $0)) }
            }

            return nil
        }
    }
}
